bundles/org.dataflowanalysis.analysis.pcm.informationflow/src/org/dataflowanalysis/analysis/pcm/informationflow/core/extraction/IFPCMExtractionStrategyPrefer.java

```diff
@@ -0,0 +1,116 @@
+
+	public IFPCMExtractionStrategyPrefer(ResourceProvider resourceProvider) {
+		super(resourceProvider);
+	}
+
+	@Override
+	protected List<ConfidentialityVariableCharacterisation> calculateResultingConfidentialityVaraibleCharacterisations(
+			List<ConfidentialityVariableCharacterisation> calculatedCharacterisations,
+			List<ConfidentialityVariableCharacterisation> definedCharacterisations,
+			Optional<DataFlowVariable> optionalSecurityContext) {
+
+		List<ConfidentialityVariableCharacterisation> definedLatticeChars = new ArrayList<>();
+		List<ConfidentialityVariableCharacterisation> definedNonLatticeChars = new ArrayList<>();
+		for (var definedChar : definedCharacterisations) {
+			LhsEnumCharacteristicReference lhsDefinedChar = (LhsEnumCharacteristicReference) definedChar.getLhs();
+
+			Literal definedLiteral = lhsDefinedChar.getLiteral();
+			EnumCharacteristicType definedCharacteristicType = (EnumCharacteristicType) lhsDefinedChar
+					.getCharacteristicType();
+
+			if (definedCharacteristicType == null) {
+				List<EnumCharacteristicType> nonLatticCharacteristicTypes = IFPCMDataDictionaryUtils
+						.getAllEnumCharacteristicTypesExceptLattice(getResourceProvider());
+				EnumCharacteristicType latticeCharacteristicType = IFPCMDataDictionaryUtils
+						.getLatticeCharacteristicType(getResourceProvider());
+
+				var resolvedNonLatticeConfChars = IFConfidentialityVariableCharacterisationUtils
+						.resolveCharacteristicTypeWildcard(definedChar, nonLatticCharacteristicTypes);
+				definedNonLatticeChars.addAll(resolvedNonLatticeConfChars);
+				var resolvedLatticeChar = IFConfidentialityVariableCharacterisationUtils.resolveWildcards(definedChar,
+						List.of(latticeCharacteristicType));
+				definedLatticeChars.addAll(resolvedLatticeChar);
+			} else if (definedLiteral == null) {
+				var resolvedChars = IFConfidentialityVariableCharacterisationUtils.resolveLiteralWildcard(definedChar);
+				if (definedCharacteristicType.getType().equals(getLattice())) {
+					definedLatticeChars.addAll(resolvedChars);
+				} else {
+					definedNonLatticeChars.addAll(resolvedChars);
+				}
+			} else if (lhsDefinedChar.getLiteral().getEnum().equals(getLattice())) {
+				definedLatticeChars.add(definedChar);
+			} else {
+				definedNonLatticeChars.add(definedChar);
+			}
+		}
+
+		checkOnlyOneCvcForEachLevel(definedLatticeChars);
+
+		List<ConfidentialityVariableCharacterisation> resultingCharacterisations = new ArrayList<>(
+				definedNonLatticeChars);
+
+		if (definedLatticeChars.isEmpty()) {
+			resultingCharacterisations.addAll(calculatedCharacterisations);
+		} else if (optionalSecurityContext.isEmpty()) {
+			resultingCharacterisations.addAll(definedCharacterisations);
+		} else {
+			resultingCharacterisations.addAll(
+					modifyResultingCvcsWithSecurityContext(definedCharacterisations, optionalSecurityContext.get()));
+		}
+
+		return resultingCharacterisations;
+	}
```

Comment:
Could you split up this method a little? Maybe lines 74ff. into one method and the loop into one?

Reply:
Changed